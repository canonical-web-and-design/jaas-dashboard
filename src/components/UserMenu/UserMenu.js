import React, { useEffect } from "react";
import { useSelector } from "react-redux";
import { Link } from "react-router-dom";
import classNames from "classnames";
import { useDispatch, useStore } from "react-redux";
import { getActiveUserTag } from "app/selectors";

import { logOut } from "app/actions";
import useAnalytics from "hooks/useAnalytics";
import { extractOwnerName } from "app/utils";

import { userMenuActive } from "ui/actions";
import { isUserMenuActive } from "ui/selectors";

import "./_user-menu.scss";

const UserMenu = () => {
  const sendAnalytics = useAnalytics();
  const dispatch = useDispatch();
  const getState = useStore().getState;
  const activeUser = useSelector(getActiveUserTag);
  const isActive = useSelector(isUserMenuActive) || false;

  useEffect(() => {
    if (isActive) {
      sendAnalytics({
        category: "User",
        action: "Opened user menu"
      });
    }
  }, [isActive, sendAnalytics]);

  return (
    <div
      className={classNames("user-menu", {
        "is-active": isActive
      })}
    >
      <>
        <div
          className="user-menu__header"
          onClick={() => dispatch(userMenuActive(!isActive))}
        >
          <i className="p-icon--user"></i>
          <span className="user-menu__name">
            {activeUser ? extractOwnerName(activeUser) : ""}
          </span>
          <i className="p-icon--contextual-menu"></i>
        </div>
        <ul className="user-menu__options">
          <li>
            <Link className="user-menu__link" to="/settings">
              Settings
            </Link>
          </li>
          <li>
            <Link
              className="user-menu__link"
              to="/"
              onClick={() => dispatch(logOut(getState))}
            >
              Log out
            </Link>
          </li>
        </ul>
      </>
    </div>
  );
};

export default UserMenu;
