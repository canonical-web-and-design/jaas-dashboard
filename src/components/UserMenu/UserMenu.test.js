import React from "react";
import { BrowserRouter as Router } from "react-router-dom";
import configureStore from "redux-mock-store";
import { Provider } from "react-redux";
import { mount } from "enzyme";
import dataDump from "testing/complete-redux-store-dump";

import UserMenu from "./UserMenu";

const mockStore = configureStore([]);
describe("User Icon", () => {
  it("renders without crashing and matches snapshot", () => {
    const store = mockStore(dataDump);
    const wrapper = mount(
      <Provider store={store}>
        <Router>
          <UserMenu />
        </Router>
      </Provider>
    );
    expect(wrapper.find(".user-menu")).toMatchSnapshot();
  });

  it("toggles drop-down panel", () => {
    const store = mockStore(dataDump);
    const wrapper = mount(
      <Provider store={store}>
        <Router>
          <UserMenu />
        </Router>
      </Provider>
    );

    const user = ".user-menu";
    const userOptionsToggle = ".user-menu__header";

    expect(wrapper.find(user).hasClass("is-expanded")).toEqual(false);
    wrapper.find(userOptionsToggle).simulate("click");
    expect(wrapper.find(user).hasClass("is-expanded")).toEqual(true);
    wrapper.find(userOptionsToggle).simulate("click");
    expect(wrapper.find(user).hasClass("is-expanded")).toEqual(false);
  });
});
