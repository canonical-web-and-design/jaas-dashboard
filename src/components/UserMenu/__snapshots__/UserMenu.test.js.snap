// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Icon renders without crashing and matches snapshot 1`] = `
<div
  className="user-menu"
>
  <div
    className="user-menu__header"
    onClick={[Function]}
  >
    <i
      className="p-icon--user"
    />
    <span
      className="user-menu__name"
    >
      Account
    </span>
    <i
      className="p-icon--contextual-menu"
    />
  </div>
  <ul
    className="user-menu__options"
  >
    <li>
      <Link
        className="user-menu__link"
        to="/settings"
      >
        <LinkAnchor
          className="user-menu__link"
          href="/settings"
          navigate={[Function]}
        >
          <a
            className="user-menu__link"
            href="/settings"
            onClick={[Function]}
          >
            Settings
          </a>
        </LinkAnchor>
      </Link>
    </li>
    <li>
      <Link
        className="user-menu__link"
        onClick={[Function]}
        to="/"
      >
        <LinkAnchor
          className="user-menu__link"
          href="/"
          navigate={[Function]}
          onClick={[Function]}
        >
          <a
            className="user-menu__link"
            href="/"
            onClick={[Function]}
          >
            Log out
          </a>
        </LinkAnchor>
      </Link>
    </li>
  </ul>
</div>
`;
